src/Compilers/VisualBasic/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.vb

```diff
@@ -598,47 +598,60 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Retargeting
             End Function
 
             Friend Function RetargetModifiers(oldModifiers As ImmutableArray(Of CustomModifier), ByRef modifiersHaveChanged As Boolean) As ImmutableArray(Of CustomModifier)
-                Dim i As Integer
-                Dim count As Integer = oldModifiers.Length
                 modifiersHaveChanged = False
+                Dim newModifiers As ArrayBuilder(Of CustomModifier) = Nothing
 
-                If count <> 0 Then
-                    Dim newModifiers As CustomModifier() = New CustomModifier(count - 1) {}
+                For i As Integer = 0 To oldModifiers.Length - 1 Step 1
+                    Dim newModifier As NamedTypeSymbol = Retarget(DirectCast(oldModifiers(i).Modifier, NamedTypeSymbol), RetargetOptions.RetargetPrimitiveTypesByName) ' should be retargeted by name
 
-                    For i = 0 To count - 1 Step 1
-                        Dim newModifier As NamedTypeSymbol = Retarget(DirectCast(oldModifiers(i).Modifier, NamedTypeSymbol), RetargetOptions.RetargetPrimitiveTypesByName) ' should be retargeted by name
+                    If Not newModifier.Equals(oldModifiers(i).Modifier) Then
+                        modifiersHaveChanged = True
 
-                        If Not newModifier.Equals(oldModifiers(i).Modifier) Then
-                            modifiersHaveChanged = True
-                            newModifiers(i) = If(oldModifiers(i).IsOptional,
-                                                VisualBasicCustomModifier.CreateOptional(newModifier),
-                                                VisualBasicCustomModifier.CreateRequired(newModifier))
-                        Else
-                            newModifiers(i) = oldModifiers(i)
+                        If newModifiers Is Nothing Then
+                            newModifiers = ArrayBuilder(Of CustomModifier).GetInstance(oldModifiers.Length)
+                            newModifiers.AddRange(oldModifiers, i)
                         End If
-                    Next
 
-                    Return newModifiers.AsImmutableOrNull()
-                End If
+                        newModifiers.Add(If(oldModifiers(i).IsOptional,
+                                                VisualBasicCustomModifier.CreateOptional(newModifier),
+                                                VisualBasicCustomModifier.CreateRequired(newModifier)))
+                    ElseIf newModifiers IsNot Nothing Then
+                        newModifiers.Add(oldModifiers(i))
+                    End If
+                Next
 
-                Return oldModifiers
+                Debug.Assert(newModifiers Is Nothing OrElse newModifiers.Count = oldModifiers.Length)
+                Debug.Assert((newModifiers IsNot Nothing) = modifiersHaveChanged)
```

Comment:
Same comment as C#: consider assigning `modifiersHaveChanged = (newModifiers IsNot Nothing)` here and removing assert.